### PR TITLE
[CALCITE-6037] The function category of ARRAY/EXTRACT_VALUE/XML_TRANSFORM/EXTRACT_XML/EXISTSNODE is incorrect

### PR DESCRIPTION
**Pull request by chucheng92**
https://issues.apache.org/jira/browse/CALCITE-6037

the definition of function category in ANSI-SQL(not very sure this is the sql standard, but the implementation of calcite is similar to it): 
https://www.oreilly.com/library/view/sql-in-a/9780596155322/ch04s04.html.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -532,26 +532,30 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
   public static final SqlFunction EXTRACT_VALUE =
       SqlBasicFunction.create("EXTRACTVALUE",
           ReturnTypes.VARCHAR_2000.andThen(SqlTypeTransforms.FORCE_NULLABLE),
-          OperandTypes.STRING_STRING);
+          OperandTypes.STRING_STRING,
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:535]: image

I find that EXTRACT_VALUE function would return the numeric value in some Scene.

I don't know whether it is appropriate to divide it into a String type function. I saw that it was divided into numeric function before.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 2]
```diff
@@ -1075,7 +1079,8 @@ private static RelDataType arrayReturnType(SqlOperatorBinding opBinding) {
   public static final SqlFunction ARRAY =
       SqlBasicFunction.create("ARRAY",
           SqlLibraryOperators::arrayReturnType,
-          OperandTypes.SAME_VARIADIC);
+          OperandTypes.SAME_VARIADIC,
+          SqlFunctionCategory.SYSTEM);
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1083]: what is meaning of System function?And why array function belong to system function.

**Comment by LakeShen** [comment]
Hi @chucheng92 ,Thank you for opening this PR, left a couple of comments!

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (chucheng92) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:535]: @LakeShen I get your point. Usually we classify functions according to the operands (in some cases, they can also be determined according to the return value type), which is not clearly specified by calcite. So we need to take a comprehensive view. These functions actually process XML. In database systems such as MySql or Oracle, they are divided into XML functions. The closest type in calcite is String.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (chucheng92) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1083]: current the function type of ARRAY is Numeric currently. Obviously not right. The closest type in calcite is System. it means built-in or system-in. we can see many other functions if not belongs to timedate/string/numeric will let it be System.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:535]: I have a meta-question: what is this field used for?
It doesn't seem to affect any of the tests.
Clearly, the meaning of this field is ambiguous too - it's not clear what the right choice is in all cases.
Can we just remove it completely?

Comment (chucheng92) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:535]: Hi, @mihaibudiu. thanks for reviewing. good question. 

Frankly speaking, `SqlFunctionCategory` is a very loose field now. In fact, there is no strict verification during lookup, so no test case covers it. However, it is currently widely used and cannot be deleted for subsequent expansion. We can refer to the comments of the current `lookupOperatorOverloads` and use it immediately when we need it.

https://github.com/apache/calcite/blob/597b53dbb075ab72700d40f66db81d2ac9c182b8/core/src/main/java/org/apache/calcite/sql/util/ReflectiveSqlOperatorTable.java#L94

**Review by JiajunBernoulli — APPROVED** [review]

**Comment by JiajunBernoulli** [comment]
For me, this is an improvement.

If there is no objection, I will merge it.

**Comment by chucheng92** [comment]
> For me, this is an improvement.
> 
> If there is no objection, I will merge it.

thanks for approve + 1.

**Pull request merged**